pkg/crd/generator/generator.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generator
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	extensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/gengo/args"
+	crdutil "sigs.k8s.io/controller-tools/pkg/crd/util"
+	"sigs.k8s.io/controller-tools/pkg/internal/codegen"
+	"sigs.k8s.io/controller-tools/pkg/internal/codegen/parse"
+	"sigs.k8s.io/controller-tools/pkg/util"
+)
+
+// Generator search API resource definition files,
+// and generate CRD install files accordingly.
+type Generator struct {
+	RootPath          string
+	Domain            string
+	Namespace         string
+	SkipMapValidation bool
+}
+
+// ValidateAndInitFields validate and init generator fields.
+func (c *Generator) ValidateAndInitFields() error {
+	var err error
+
+	if len(c.RootPath) == 0 {
+		// Take current path as root path if not specified.
+		c.RootPath, err = os.Getwd()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Validate root path is under go src path
+	if !crdutil.IsUnderGoSrcPath(c.RootPath) {
```

Comment:
Wondering why is this a requirement ?

Reply:
As both old kubebuilder and new controller-tools command has the limitation that command need to be run under go src path, the generated api resource path `pkg/apis` is thought to be under go src path.

Reply:
ok, its fine for now.